mne_sandbox/connectivity/cfc.py

```diff
@@ -0,0 +1,623 @@
+                        func = getattr(ppac, i_pac_func)
+                        pac[ix_func, iep, iix, i_f_pair, itime] = func(
+                            ep_ph[i_ix_ph], ep_am[i_ix_am],
+                            f_phase, f_amp, filterfn=False)
+            pbar.update_with_increment_value(1)
+    if pac.shape[0] == 1:
+        pac = pac[0]
+    if return_data:
+        return pac, freq_pac, data_ph, data_am
+    else:
+        return pac, freq_pac
+
+
+def _raw_to_epochs_mne(raw, events, tmin, tmax):
+    """Convert Raw data to Epochs w/ some time checks."""
+    events = np.atleast_1d(events)
+    if events.ndim == 1:
+        events = np.vstack([events, np.zeros_like(events),
+                            np.ones_like(events)]).T
+    if events.ndim != 2:
+        raise ValueError('events have incorrect number of dimensions')
+    if events.shape[-1] != 3:
+        raise ValueError('events have incorrect number of columns')
+    # Convert to Epochs using the event times
+    tmin_all = np.min(tmin)
+    tmax_all = np.max(tmax) + (1. / raw.info['sfreq'])
+    return mne.Epochs(raw, events, tmin=tmin_all, tmax=tmax_all, preload=True,
+                      baseline=None)
+
+
+def _pre_filter_ph_am(data, sfreq, ixs, f_ph, f_am, n_cycles=3,
+                      hi_phase=False, scale_amp_func=None, kws_filt=None):
+    """Filter for phase/amp only once for each channel."""
+    from ..externals.pacpy.pac import _range_sanity
+    from scipy.signal import hilbert
+
+    kws_filt = dict() if kws_filt is None else kws_filt
+    ix_ph = np.atleast_1d(np.unique(ixs[:, 0]))
+    ix_am = np.atleast_1d(np.unique(ixs[:, 1]))
+    n_times = data.shape[-1]
+
+    # Filter for lo-freq phase
+    for i_f_ph in f_ph:
+        _range_sanity(i_f_ph, f_am[0])
+    for i_f_am in f_am:
+        _range_sanity(f_ph[0], i_f_am)
+    data_ph = data[ix_ph, :]
+    # Output will be (n_chan, n_freqs, n_times)
+    out_ph = np.zeros([data_ph.shape[0], f_ph.shape[0], data_ph.shape[1]])
+    for ii in range(data_ph.shape[0]):
+        for jj, i_f_ph in enumerate(f_ph):
+            out_ph[ii, jj] = _band_pass_pac(data_ph[ii], i_f_ph, sfreq,
+                                            n_cycles=n_cycles)
+    # Now calculate phase w/ Hilbert
+    N_hil = int(2 ** np.ceil(np.log2(out_ph.shape[-1])))
+    out_ph = np.angle(hilbert(out_ph, N=N_hil)[..., :n_times])
+    ix_map_ph = dict((ix, i) for i, ix in enumerate(ix_ph))
+
+    # Filter for hi-freq amplitude
+    data_am = data[ix_am, :]
+    out_am = np.zeros([data_am.shape[0], f_am.shape[0], data_am.shape[1]])
+    for ii in range(data_am.shape[0]):
+        for jj, i_f_am in enumerate(f_am):
+            out_am[ii, jj] = _band_pass_pac(data_am[ii], i_f_am, sfreq,
+                                            n_cycles=n_cycles)
+        N_hil = int(2 ** np.ceil(np.log2(out_am.shape[-1])))
+        out_am = np.abs(hilbert(out_am, N=N_hil)[..., :n_times])
+
+    if hi_phase is True:
```

Comment:
I think we actually *want* double filtering. This is specifically for the `plv` metric, which is actually calculating `phase-phase` coupling between the phase of the low-freq channel, and the phase of the low-freq component of the amplitude of the high-freq component of the other channel. So it's basically doing `corr(phase(band_lo(sig_a)), phase(band_lo(amp(band_hi(sig_b)))))`. At least, that's what I think it's supposed to do...correct if I'm wrong!